File: ios/SDKSampleApp/Shared/DemoSession.swift
import Foundation
import LiveVoiceSDK

@MainActor
final class DemoSession: ObservableObject {
    @Published private(set) var selectedPreset: DemoApiKeyPreset

    /// Drives the app-level API key picker sheet.
    @Published var isApiKeyPickerPresented = false

    /// Drives the "Use your own event" credentials editor sheet.
    @Published var isCustomCredentialsEditorPresented = false

    /// Last credentials a tester entered via "Use your own event".
    @Published private(set) var customCredentials: DemoCredentials?

    /// Set when the picker asks for the editor.
    var pendingCustomCredentialsEditor = false

    private static let customCredentialsDefaultsKey = "DemoSession.customCredentials"

    init(
        selectedPreset: DemoApiKeyPreset = DemoFixture.defaultPreset,
        joinOnInit: Bool = true
    ) {
        self.selectedPreset = selectedPreset
        self.customCredentials = Self.loadCustomCredentials()

        if joinOnInit {
            joinSelectedPreset()
        }
    }

    func select(_ preset: DemoApiKeyPreset) {
        guard preset != selectedPreset else { return }
        selectedPreset = preset
        joinSelectedPreset()
    }

    /// Called from the picker's "Use your own event" row.
    func requestCustomCredentialsEditor() {
        pendingCustomCredentialsEditor = true
        isApiKeyPickerPresented = false
    }

    /// Applies credentials entered in the editor: persists them, selects the
    /// custom event, and (re)joins. An empty password means "no password".
    func applyCustomCredentials(joinCode: String, apiKey: String, password: String) {
        let credentials = DemoCredentials(joinCode: joinCode, password: password, apiKey: apiKey)
        customCredentials = credentials
        persistCustomCredentials(credentials)
        select(DemoFixture.customPreset(from: credentials))
    }

    private func joinSelectedPreset() {
        let preset = selectedPreset
        Task {
            await LiveVoice.shared.joinEvent(
                joinCode: preset.joinCode,
                password: preset.password,
                apiKey: preset.apiKey
            )
        }
    }

    private func persistCustomCredentials(_ credentials: DemoCredentials) {
        guard let data = try? JSONEncoder().encode(credentials) else { return }
        UserDefaults.standard.set(data, forKey: Self.customCredentialsDefaultsKey)
    }

    private static func loadCustomCredentials() -> DemoCredentials? {
        guard let data = UserDefaults.standard.data(forKey: customCredentialsDefaultsKey) else {
            return nil
        }
        return try? JSONDecoder().decode(DemoCredentials.self, from: data)
    }

    static var preview: DemoSession {
        DemoSession(joinOnInit: false)
    }
}

File: android/app/src/main/java/io/livevoice/example/shared/DemoSession.kt
package io.livevoice.example.shared

import android.content.Context
import io.livevoice.sdk.android.publicApi.core.LiveVoice
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class DemoSession(
    private val scope: CoroutineScope,
    context: Context,
    initialPreset: DemoApiKeyPreset = DemoFixture.defaultPreset,
) {
    private val prefs = context.applicationContext
        .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private val _selectedPreset = MutableStateFlow(initialPreset)
    val selectedPreset: StateFlow<DemoApiKeyPreset> = _selectedPreset.asStateFlow()

    // Last credentials a tester entered via "Use your own event".
    private val _customCredentials = MutableStateFlow(loadCustomCredentials())
    val customCredentials: StateFlow<DemoCredentials?> = _customCredentials.asStateFlow()

    private var joinJob: Job? = null

    fun start() {
        joinSelectedPreset(_selectedPreset.value)
    }

    fun select(preset: DemoApiKeyPreset) {
        if (preset == _selectedPreset.value) return
        _selectedPreset.value = preset
        joinSelectedPreset(preset)
    }

    /**
     * Applies credentials entered in the editor: persists them, selects the
     * custom event, and (re)joins. An empty password means "no password".
     */
    fun applyCustomCredentials(credentials: DemoCredentials) {
        _customCredentials.value = credentials
        persistCustomCredentials(credentials)
        select(DemoFixture.customPreset(credentials))
    }

    fun stop() {
        joinJob?.cancel()
        joinJob = null
        LiveVoice.leaveEvent()
    }

    private fun joinSelectedPreset(preset: DemoApiKeyPreset) {
        joinJob?.cancel()
        LiveVoice.leaveEvent()

        joinJob = scope.launch {
            LiveVoice.joinEvent(
                joinCode = preset.joinCode,
                password = preset.password,
                apiKey = preset.apiKey
            )
        }
    }

    private fun persistCustomCredentials(credentials: DemoCredentials) {
        prefs.edit()
            .putString(KEY_JOIN_CODE, credentials.joinCode)
            .putString(KEY_PASSWORD, credentials.password)
            .putString(KEY_API_KEY, credentials.apiKey)
            .apply()
    }

    private fun loadCustomCredentials(): DemoCredentials? {
        val joinCode = prefs.getString(KEY_JOIN_CODE, null) ?: return null
        val apiKey = prefs.getString(KEY_API_KEY, null) ?: return null
        val password = prefs.getString(KEY_PASSWORD, "").orEmpty()
        return DemoCredentials(joinCode = joinCode, password = password, apiKey = apiKey)
    }

    private companion object {
        const val PREFS_NAME = "demo_session"
        const val KEY_JOIN_CODE = "custom_join_code"
        const val KEY_PASSWORD = "custom_password"
        const val KEY_API_KEY = "custom_api_key"
    }
}
